### Overriding Properties

* We can change any property which is not defined as final in **core-site.xml** or **hdfs-site.xml**.
* We can change `blocksize` as well as `replication` while copying the files. We can also change them after copying the files as well.
* We can either pass individual properties using `-D` or bunch of properties by passing xml similar to **core-site.xml** or **hdfs-site.xml** as part of `--conf`.
* Let's copy a file **/data/crime/csv/rows.csv** with default values. The file is splitted into 12 blocks with 2 copies each (as our default blocksize is 128 MB and replication factor is 2).

In [9]:
! ls -lh -r  /data/

total 52K
drwxr-xr-x  2 itversity itversity 4.0K Jan 26 05:06 sms_db
drwxr-xr-x  8 itversity itversity 4.0K Jan 26 05:06 retail_db_json
drwxr-xr-x  8 itversity itversity 4.0K Jan 26 05:06 retail_db
drwxr-xr-x  4 itversity itversity 4.0K Jan 26 05:06 nyse_all
drwxr-xr-x  2 itversity itversity 4.0K Jan 26 05:06 nyse
drwxr-xr-x  2 itversity itversity 4.0K Jan 26 05:06 lca
drwxr-xr-x 10 itversity itversity 4.0K Jan 26 05:06 hr_db
drwxr-xr-x  2 itversity itversity 4.0K Jan 26 05:06 hr
drwxr-xr-x  2 itversity itversity 4.0K Jan 26 05:06 electionresults
drwxr-xr-x  2 root      root      4.0K Jan 28 21:58 crims
drwxr-xr-x  3 itversity itversity 4.0K Jan 28 21:59 crime
drwxr-xr-x  4 itversity itversity 4.0K Jan 26 05:06 cards
-rw-r--r--  1 itversity itversity  160 Jan 26 05:06 README.md


In [10]:
%%sh

hdfs dfs -ls /public 

Found 2 items
drwxr-xr-x   - itversity supergroup          0 2024-01-27 23:11 /public/nyse_all
drwxr-xr-x   - itversity supergroup          0 2024-01-28 16:56 /public/retail_db


In [11]:
%%sh

hdfs dfs -ls /user/`whoami`/crime

Found 1 items
drwxr-xr-x   - itversity supergroup          0 2024-01-28 22:40 /user/itversity/crime/csv


In [12]:
%%sh

hdfs dfs -rm -R -skipTrash /user/`whoami`/crime

Deleted /user/itversity/crime


In [13]:
%%sh

hdfs dfs -mkdir -p /user/`whoami`/crime/csv

In [14]:
%%sh

ls -lhtr /data/crime/csv

total 1.8G
-rw-r--r-- 1 root root 1.8G Jan 28 20:57 rows.csv
-rw-r--r-- 1 root root   67 Jan 28 20:57 Crimes_-_2001_to_Present_20240128.csv:Zone.Identifier


In [15]:
%%sh

hdfs dfs -put /data/crime/csv/rows.csv /user/`whoami`/crime/csv

In [16]:
%%sh

hdfs dfs -stat %r /user/`whoami`/crime/csv/rows.csv

1


In [17]:
%%sh

hdfs dfs -stat %o /user/`whoami`/crime/csv/rows.csv

134217728


In [18]:
%%sh

hdfs dfs -stat %b /user/`whoami`/crime/csv/rows.csv

1885779202


In [19]:
%%sh

hdfs dfs -rm -R -skipTrash /user/`whoami`/crime/csv/rows.csv

Deleted /user/itversity/crime/csv/rows.csv


In [20]:
%%sh

hdfs dfs -put /data/crime/csv/rows.csv /user/`whoami`/crime/csv

In [21]:
%%sh

hdfs fsck /user/`whoami`/crime/csv

FSCK started by itversity (auth:SIMPLE) from /127.0.0.1 for path /user/itversity/crime/csv at Sun Jan 28 22:49:41 GMT 2024


Status: HEALTHY
 Number of data-nodes:	1
 Number of racks:		1
 Total dirs:			1
 Total symlinks:		0

Replicated Blocks:
 Total size:	1885779202 B
 Total files:	1
 Total blocks (validated):	15 (avg. block size 125718613 B)
 Minimally replicated blocks:	15 (100.0 %)
 Over-replicated blocks:	0 (0.0 %)
 Under-replicated blocks:	0 (0.0 %)
 Mis-replicated blocks:		0 (0.0 %)
 Default replication factor:	1
 Average block replication:	1.0
 Missing blocks:		0
 Corrupt blocks:		0
 Missing replicas:		0 (0.0 %)
 Blocks queued for replication:	0

Erasure Coded Block Groups:
 Total size:	0 B
 Total files:	0
 Total block groups (validated):	0
 Minimally erasure-coded block groups:	0
 Over-erasure-coded block groups:	0
 Under-erasure-coded block groups:	0
 Unsatisfactory placement block groups:	0
 Average block group size:	0.0
 Missing block groups:		0
 Corrupt block groups:		0
 M

Connecting to namenode via http://localhost:9870/fsck?ugi=itversity&path=%2Fuser%2Fitversity%2Fcrime%2Fcsv


In [23]:
%%sh

hdfs fsck /user/`whoami`/crime/csv -files -blocks -locations

FSCK started by itversity (auth:SIMPLE) from /127.0.0.1 for path /user/itversity/crime/csv at Sun Jan 28 22:50:34 GMT 2024

/user/itversity/crime/csv <dir>
/user/itversity/crime/csv/rows.csv 1885779202 bytes, replicated: replication=1, 15 block(s):  OK
0. BP-1463716262-172.19.0.3-1706254608807:blk_1073742486_1662 len=134217728 Live_repl=1  [DatanodeInfoWithStorage[127.0.0.1:9866,DS-16774d13-cf92-478b-9b30-c7794943f0f1,DISK]]
1. BP-1463716262-172.19.0.3-1706254608807:blk_1073742487_1663 len=134217728 Live_repl=1  [DatanodeInfoWithStorage[127.0.0.1:9866,DS-16774d13-cf92-478b-9b30-c7794943f0f1,DISK]]
2. BP-1463716262-172.19.0.3-1706254608807:blk_1073742488_1664 len=134217728 Live_repl=1  [DatanodeInfoWithStorage[127.0.0.1:9866,DS-16774d13-cf92-478b-9b30-c7794943f0f1,DISK]]
3. BP-1463716262-172.19.0.3-1706254608807:blk_1073742489_1665 len=134217728 Live_repl=1  [DatanodeInfoWithStorage[127.0.0.1:9866,DS-16774d13-cf92-478b-9b30-c7794943f0f1,DISK]]
4. BP-1463716262-172.19.0.3-1706254608807:b

Connecting to namenode via http://localhost:9870/fsck?ugi=itversity&files=1&blocks=1&locations=1&path=%2Fuser%2Fitversity%2Fcrime%2Fcsv


In [44]:
%%sh

hdfs dfs -rm -R -skipTrash /user/`whoami`/crime/csv/rows.csv

Deleted /user/itversity/crime/csv/rows.csv


In [35]:
%%sh

hdfs dfs -Ddfs.blocksize=64M -Ddfs.replication=2 -put /data/crime/csv/rows.csv /user/`whoami`/crime/csv

In [36]:
%%sh

hdfs dfs -ls /user/`whoami`/crime/csv/

Found 1 items
-rw-r--r--   2 itversity supergroup 1885779202 2024-01-28 22:55 /user/itversity/crime/csv/rows.csv


In [37]:
%%sh

hdfs fsck /user/`whoami`/crime/csv -files -blocks -locations

FSCK started by itversity (auth:SIMPLE) from /127.0.0.1 for path /user/itversity/crime/csv at Sun Jan 28 22:56:08 GMT 2024

/user/itversity/crime/csv <dir>
/user/itversity/crime/csv/rows.csv 1885779202 bytes, replicated: replication=2, 29 block(s):  Under replicated BP-1463716262-172.19.0.3-1706254608807:blk_1073742530_1706. Target Replicas is 2 but found 1 live replica(s), 0 decommissioned replica(s), 0 decommissioning replica(s).
 Under replicated BP-1463716262-172.19.0.3-1706254608807:blk_1073742531_1707. Target Replicas is 2 but found 1 live replica(s), 0 decommissioned replica(s), 0 decommissioning replica(s).
 Under replicated BP-1463716262-172.19.0.3-1706254608807:blk_1073742532_1708. Target Replicas is 2 but found 1 live replica(s), 0 decommissioned replica(s), 0 decommissioning replica(s).
 Under replicated BP-1463716262-172.19.0.3-1706254608807:blk_1073742533_1709. Target Replicas is 2 but found 1 live replica(s), 0 decommissioned replica(s), 0 decommissioning replica(s).
 Un

Connecting to namenode via http://localhost:9870/fsck?ugi=itversity&files=1&blocks=1&locations=1&path=%2Fuser%2Fitversity%2Fcrime%2Fcsv


In [39]:
%%sh

hdfs dfs -stat %r /user/`whoami`/crime/csv/rows.csv

2


In [40]:
%%sh

hdfs dfs -stat %o /user/`whoami`/crime/csv/rows.csv

67108864


In [41]:
%%sh

hdfs dfs -stat %b /user/`whoami`/crime/csv/rows.csv

1885779202


In [42]:
%%sh

ls -ltr /etc/hadoop/conf/

total 164
-rw-r--r-- 1 itversity itversity    10 Jul  6  2020 workers
-rw-r--r-- 1 itversity itversity  2697 Jul  6  2020 ssl-server.xml.example
-rw-r--r-- 1 itversity itversity  2316 Jul  6  2020 ssl-client.xml.example
drwxr-xr-x 2 itversity itversity  4096 Jul  6  2020 shellprofile.d
-rw-r--r-- 1 itversity itversity 14032 Jul  6  2020 log4j.properties
-rw-r--r-- 1 itversity itversity  3414 Jul  6  2020 hadoop-user-functions.sh.example
-rw-r--r-- 1 itversity itversity 11765 Jul  6  2020 hadoop-policy.xml
-rw-r--r-- 1 itversity itversity  3321 Jul  6  2020 hadoop-metrics2.properties
-rw-r--r-- 1 itversity itversity  3999 Jul  6  2020 hadoop-env.cmd
-rw-r--r-- 1 itversity itversity   682 Jul  6  2020 kms-site.xml
-rw-r--r-- 1 itversity itversity  1860 Jul  6  2020 kms-log4j.properties
-rw-r--r-- 1 itversity itversity  1351 Jul  6  2020 kms-env.sh
-rw-r--r-- 1 itversity itversity  3518 Jul  6  2020 kms-acls.xml
-rw-r--r-- 1 itversity itversity  2681 Jul  6  2020 user_ec_policies.xml.temp

In [43]:
%%sh 

cat /etc/hadoop/conf/hdfs-site.xml

<configuration>
    <property>
        <name>dfs.namenode.name.dir</name>
        <value>/opt/hadoop/dfs/name</value>
    </property>
    <property>
        <name>dfs.namenode.checkpoint.dir</name>
        <value>/opt/hadoop/dfs/namesecondary</value>
    </property>
    <property>
        <name>dfs.datanode.data.dir</name>
        <value>/opt/hadoop/dfs/data</value>
    </property>
    <property>
        <name>dfs.replication</name>
        <value>1</value>
    </property>
</configuration>


*************************************************************

*** configuration file example

*************************************************************

  <configuration>
    
    <property>
      <name>dfs.block.access.token.enable</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.blockreport.initialDelay</name>
      <value>120</value>
    </property>
    
    <property>
      <name>dfs.blocksize</name>
      <value>134217728</value>
    </property>
    
    <property>
      <name>dfs.client.read.shortcircuit</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.client.read.shortcircuit.streams.cache.size</name>
      <value>4096</value>
    </property>
    
    <property>
      <name>dfs.client.retry.policy.enabled</name>
      <value>false</value>
    </property>
    
    <property>
      <name>dfs.cluster.administrators</name>
      <value> hdfs</value>
    </property>
    
    <property>
      <name>dfs.content-summary.limit</name>
      <value>5000</value>
    </property>
    
    <property>
      <name>dfs.datanode.address</name>
      <value>0.0.0.0:50010</value>
    </property>
    
    <property>
      <name>dfs.datanode.balance.bandwidthPerSec</name>
      <value>6250000</value>
    </property>
    
    <property>
      <name>dfs.datanode.data.dir</name>
      <value>/hdp01/hadoop/hdfs/data,/hdp02/hadoop/hdfs/data</value>
    </property>
    
    <property>
      <name>dfs.datanode.data.dir.perm</name>
      <value>750</value>
    </property>
    
    <property>
      <name>dfs.datanode.du.reserved</name>
      <value>2563350016</value>
    </property>
    
    <property>
      <name>dfs.datanode.failed.volumes.tolerated</name>
      <value>0</value>
    </property>
    
    <property>
      <name>dfs.datanode.http.address</name>
      <value>0.0.0.0:50075</value>
    </property>
    
    <property>
      <name>dfs.datanode.https.address</name>
      <value>0.0.0.0:50475</value>
    </property>
    
    <property>
      <name>dfs.datanode.ipc.address</name>
      <value>0.0.0.0:8010</value>
    </property>
    
    <property>
      <name>dfs.datanode.max.transfer.threads</name>
      <value>4096</value>
    </property>
    
    <property>
      <name>dfs.domain.socket.path</name>
      <value>/var/lib/hadoop-hdfs/dn_socket</value>
    </property>
    
    <property>
      <name>dfs.encrypt.data.transfer.cipher.suites</name>
      <value>AES/CTR/NoPadding</value>
    </property>
    
    <property>
      <name>dfs.heartbeat.interval</name>
      <value>3</value>
    </property>
    
    <property>
      <name>dfs.hosts.exclude</name>
      <value>/etc/hadoop/conf/dfs.exclude</value>
    </property>
    
    <property>
      <name>dfs.http.policy</name>
      <value>HTTP_ONLY</value>
    </property>
    
    <property>
      <name>dfs.https.port</name>
      <value>50470</value>
    </property>
    
    <property>
      <name>dfs.journalnode.edits.dir</name>
      <value>/hadoop/hdfs/journalnode</value>
    </property>
    
    <property>
      <name>dfs.journalnode.http-address</name>
      <value>0.0.0.0:8480</value>
    </property>
    
    <property>
      <name>dfs.journalnode.https-address</name>
      <value>0.0.0.0:8481</value>
    </property>
    
    <property>
      <name>dfs.namenode.accesstime.precision</name>
      <value>3600000</value>
    </property>
    
    <property>
      <name>dfs.namenode.audit.log.async</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.namenode.avoid.read.stale.datanode</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.namenode.avoid.write.stale.datanode</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.namenode.checkpoint.dir</name>
      <value>/var/hadoop/hdfs/namesecondary</value>
    </property>
    
    <property>
      <name>dfs.namenode.checkpoint.edits.dir</name>
      <value>${dfs.namenode.checkpoint.dir}</value>
    </property>
    
    <property>
      <name>dfs.namenode.checkpoint.period</name>
      <value>21600</value>
    </property>
    
    <property>
      <name>dfs.namenode.checkpoint.txns</name>
      <value>1000000</value>
    </property>
    
    <property>
      <name>dfs.namenode.fslock.fair</name>
      <value>false</value>
    </property>
    
    <property>
      <name>dfs.namenode.handler.count</name>
      <value>200</value>
    </property>
    
    <property>
      <name>dfs.namenode.http-address</name>
      <value>172.16.1.101:50070</value>
    </property>
    
    <property>
      <name>dfs.namenode.https-address</name>
      <value>nn01.itversity.com:50470</value>
    </property>
    
    <property>
      <name>dfs.namenode.name.dir</name>
      <value>/hdp01/hadoop/hdfs/namenode,/hdp02/hadoop/hdfs/namenode</value>
    </property>
    
    <property>
      <name>dfs.namenode.name.dir.restore</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.namenode.rpc-address</name>
      <value>nn01.itversity.com:8020</value>
    </property>
    
    <property>
      <name>dfs.namenode.safemode.threshold-pct</name>
      <value>1</value>
    </property>
    
    <property>
      <name>dfs.namenode.secondary.http-address</name>
      <value>nn02.itversity.com:50090</value>
    </property>
    
    <property>
      <name>dfs.namenode.stale.datanode.interval</name>
      <value>30000</value>
    </property>
    
    <property>
      <name>dfs.namenode.startup.delay.block.deletion.sec</name>
      <value>3600</value>
    </property>
    
    <property>
      <name>dfs.namenode.write.stale.datanode.ratio</name>
      <value>1.0f</value>
    </property>
    
    <property>
      <name>dfs.permissions.enabled</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.permissions.superusergroup</name>
      <value>hdfs</value>
    </property>
    
    <property>
      <name>dfs.replication</name>
      <value>2</value>
    </property>
    
    <property>
      <name>dfs.replication.max</name>
      <value>50</value>
    </property>
    
    <property>
      <name>dfs.support.append</name>
      <value>true</value>
    </property>
    
    <property>
      <name>dfs.webhdfs.enabled</name>
      <value>true</value>
    </property>
    
    <property>
      <name>fs.permissions.umask-mode</name>
      <value>022</value>
    </property>
    
    <property>
      <name>nfs.exports.allowed.hosts</name>
      <value>* rw</value>
    </property>
    
    <property>
      <name>nfs.file.dump.dir</name>
      <value>/tmp/.hdfs-nfs</value>
    </property>
    
  </configuration>